cirq/ops/arithmetic_operation.py

```diff
@@ -0,0 +1,248 @@
+        ...    Add(target_register=cirq.LineQubit.range(4),
+        ...        input_register=cirq.LineQubit.range(4, 8)),
+        ...
+        ...    cirq.measure(*cirq.LineQubit.range(4, 8), key='after:in'),
+        ...    cirq.measure(*cirq.LineQubit.range(4), key='after:out'),
+        ... )
+        >>> cirq.sample(c).data
+           before:in  before:out  after:in  after:out
+        0          2           3         2          5
+    """
+
+    @abc.abstractmethod
+    def registers(self) -> Sequence[Union[int, Sequence['cirq.Qid']]]:
+        """The data acted upon by the arithmetic operation.
+
+        Each register in the list can either be a classical constant (an `int`),
+        or else a list of qubits/qudits (a `List[cirq.Qid]`). Registers that
+        are set to a classical constant must not be mutated by the arithmetic
+        operation (their value must remain fixed when passed to `apply`).
+
+        Registers are big endian. The first qubit is the most significant, the
+        last qubit is the 1s qubit, the before last qubit is the 2s qubit, etc.
+
+        Returns:
+            A list of constants and qubit groups that the operation will act
+            upon.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def with_registers(self, *new_registers: Union[int, Sequence['cirq.Qid']]
+                      ) -> 'cirq.ArithmeticOperation':
+        """Returns the same operation targeting different registers.
+
+        Args:
+            new_registers: The new values that should be returned by the
+                `registers` method.
+
+        Returns:
+            An instance of the same kind of operation, but acting on different
+            registers.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def apply(self, *register_values: int) -> Union[int, Iterable[int]]:
+        """Returns the result of the operation operating on classical values.
+
+        For example, an addition takes two values (the target and the source),
+        adds the source into the target, then returns the target and source
+        as the new register values.
+
+        The `apply` method is permitted to be sloppy in three ways:
+
+        1. The `apply` method is permitted to return values that have more bits
+            than the registers they will be stored into. The extra bits are
+            simply dropped. For example, if the value 5 is returned for a 2
+            qubit register then 5 % 2**2 = 1 will be used instead. Negative
+            values are also permitted. For example, for a 3 qubit register the
+            value -2 becomes -2 % 2**3 = 6.
+        2. When the value of the last `k` registers is not changed by the
+            operation, the `apply` method is permitted to omit these values
+            from the result. That is to say, when the length of the output is
+            less than the length of the input, it is padded up to the intended
```

Comment:
I'm mildly confused about the description of this feature; especially where it talks about "padding".

Is this an example: If I had a "summation" gate that takes `k` registers and puts the sum in the first register, my `apply` method would only return one number. Where is padding happening? How is this point (2) different than the next one (3)?

Reply:
If you have two modified registers and three input registers, you can return a pair of values instead of a quintuplet of values. It's true that this can be seen as a generalization of (3), but the in the code they refer to separate steps.